### PR TITLE
[SimplifyCFG] Deduce paths unreachable if they cause div/rem UB

### PR DESCRIPTION
**Pull request by goldsteinn**
- **[SimplifyCFG] Add tests for deducing paths unreachable if they cause div/rem UB; NFC**
- **[SimplifyCFG] Deduce paths unreachable if they cause div/rem UB**

Same we way mark a path unreachable if it may cause a nullptr
dereference, div/rem by zero or signed div/rem of INT_MIN by -1 cause
immediate UB.

### CODE REVIEW

**Review by nikic — COMMENTED** [review]
I think the division by zero part of this is reasonable, but handling SignedMin / -1 is an unnecessary complication, especially as it adds another constant root. It's not like `SignedMin / x` is common code.

**Comment by goldsteinn** [comment]
> I think the division by zero part of this is reasonable, but handling SignedMin / -1 is an unnecessary complication, especially as it adds another constant root. It's not like `SignedMin / x` is common code.

Dropped the INT_MIN / -1 logic.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 1]
```diff
@@ -7882,6 +7882,14 @@ static bool passingValueIsAlwaysUndefined(Value *V, Instruction *I, bool PtrValu
...
       case Instruction::Invoke:
         return true;
+      case Instruction::UDiv:
+      case Instruction::URem:
+        // Note: signed div/rem of INT_MIN / -1 is also immediate UB, not
+        // implemented to avoid code complexity as it is unclear how useful such
+        // logic is.
+      case Instruction::SDiv:
+      case Instruction::SRem:
+        return C->isNullValue();
```
Comment (nikic) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7892]: Don't think you need this check? Division by undef is also UB (if undef is chosen as 0).

Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7892]: Don't we need to worry about undef being chosen on non-zero?

Comment (nikic) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp:7892]: As undef is chosen per-use, we can always chose zero here, even if another use is non-zero.

Comment (goldsteinn) [reply llvm/lib/Transforms/Utils/SimplifyCFG.cpp]: Done :)

**Review by dtcxzyw — APPROVED** [review]
LG

File: llvm/test/Transforms/SimplifyCFG/UnreachableEliminate.ll [hunk 2]
```diff
@@ -918,6 +918,280 @@ bb5:                                              ; preds = %bb3, %bb
   ret i32 %i7
 }
 
+declare void @side.effect()
+declare i8 @get.i8()
+
+define i8 @udiv_by_zero(i8 noundef %x, i8 noundef %i, i8 noundef %v) {
```
Comment (nikic) [inline llvm/test/Transforms/SimplifyCFG/UnreachableEliminate.ll:924]: Are the noundefs relevant?

**Review by nikic — APPROVED** [review]
LGTM, but please fix the code formatting.

File: llvm/test/Transforms/SimplifyCFG/UnreachableEliminate.ll (hunk 2 above)
Comment (goldsteinn) [reply llvm/test/Transforms/SimplifyCFG/UnreachableEliminate.ll:924]: No

**Comment by bwendling** [comment]
This patch is causing major issues with Linux where it silently generates bad code. In the reported case, it's falling into the next function.

I don't think this patch works, because unlike floating point math, there's no way to tell the compiler to accept divide by zero or shifting by a negative number.

**Pull request closed without merging**